File: ml/dataset/build_user_tensor.py
import math

def build_user_tensor(context, embedding_length):

    result = []

    pad = [0.0] * (embedding_length + 2)

    for i in range(10):

        event = context[i] if i < len(context) else None

        if event is None:
            result.append(pad)
            continue

        type = [0.0, 1.0] if event["eventType"] == 1 else [1.0, 0.0]

        result.append([
            *event["songEmbedding"],
            *type,
            math.log1p(event["timeDeltaHours"]) / 5.0
        ])

    return result

File: server/services/recommendations/dataset/buildUserTensor.js
export function buildUserTensor(context, embeddingLength) {

  const result = []
  const pad = Array(embeddingLength + 2).fill(0)

  for (let i = 0; i < 10; i++) {

    const event = context[i]

    if (!event) {
      result.push(pad)
      continue
    }

    const type = event.eventType === 1 ? [0, 1] : [1, 0]

    result.push([
      ...event.songEmbedding,
      ...type,
      Math.log1p(event.timeDeltaHours) / 5
    ])
  }

  return result
}